compiler/rustc_ty_utils/src/layout.rs

```diff
@@ -304,18 +309,22 @@ fn layout_of_uncached<'tcx>(
 
             // SIMD vectors with zero fields are not supported.
             // (should be caught by typeck)
-            if def.non_enum_variant().fields.is_empty() {
-                tcx.sess.fatal(&format!("monomorphising SIMD type `{}` of zero length", ty));
+            if fields.is_empty() {
+                tcx.sess.emit_fatal(ZeroLengthSimdType { ty })
             }
 
             // Type of the first ADT field:
-            let f0_ty = def.non_enum_variant().fields[FieldIdx::from_u32(0)].ty(tcx, substs);
+            let f0_ty = fields[FieldIdx::from_u32(0)].ty(tcx, substs);
 
             // Heterogeneous SIMD vectors are not supported:
             // (should be caught by typeck)
-            for fi in &def.non_enum_variant().fields {
+            for fi in fields {
                 if fi.ty(tcx, substs) != f0_ty {
-                    tcx.sess.fatal(&format!("monomorphising heterogeneous SIMD type `{}`", ty));
+                    tcx.sess.delay_span_bug(
```

Comment:
This doesn't just migrate the diagnostic, but also changes the behavior. Do you have a test that triggers this?

Reply:
`error-codes\E0076.rs` and `simd\type-len.rs` show the error that's emitted before this.

Reply:
Sure but I was asking for something that actually triggered the fatal error here. Here is a UI test that enters this codepath. Could you add it?

```
#![feature(repr_simd)]

#[repr(simd)]
struct I64F64(i64, f64);

static X: I64F64 = I64F64(1, 2.0);
```